### PR TITLE
[maxtext] Add block-diffusion CFT and SFT

### PR DESCRIPTION
**Pull request by ethannnnnn**
## Motivation

#4776 added the opt-in block-diffusion pre-training path. MaxText now needs a
focused text CFT and completion-only SFT integration before rollout,
distillation, replay, or RL behavior is introduced.

## Dependencies

This change follows and depends on:

- #4776 for MaxText block-diffusion pre-training; and
- google/tunix#1891 for target-aligned diffusion SFT batches and weighted
  trainer loss.

Both dependencies are merged. Current MaxText pins upstream Tunix commit
`a8d70582f1e2f1fb65973210989e0e148b5ef7ad`, a descendant of the Copybara
integration of google/tunix#1891 (`bd1b0f47606a8bfb5158fe17084503c16bb4a62e`).

## Scope

This PR adds only:

- opt-in Hugging Face text CFT and completion-only SFT for
  `training_objective=block_diffusion`;
- clean targets, role-derived completion eligibility, corruption masks, and
  explicit target-loss weights through the SFT pipeline;
- a MaxText-to-Tunix adapter for target-aligned diffusion batches and logits;
- weighted training and preaveraged evaluation metrics;
- safe trainer routing for diffusion gradient accumulation and internal neural
  network metrics; and
- fail-closed validation for prompt/completion layouts that would expose a
  later prompt to an earlier completion inside one bidirectional diffusion
  block.

It does not add DPO, multimodal or audio training, rollout, serving,
distillation, replay, or RL. Block-diffusion training continues to require an
unpacked Hugging Face text pipeline.

## Training contract

The pre-training logit and canvas contracts remain unchanged:

- `same_position` logits with an `all_masked` canvas; and
- `shifted` logits with a `seed_and_mask` canvas.

CFT corrupts and supervises all valid target positions. Completion-only SFT
keeps prompt tokens as clean context and corrupts and supervises completion
positions only. Because attention is bidirectional within a diffusion block,
each block must contain a prompt prefix followed by a completion suffix. A
later prompt may start at the next block boundary; an assistant-to-user
transition inside one block fails with an actionable error instead of leaking
future prompt text.

The Tunix adapter requires non-None position, segmentation, completion,
corruption, and loss-mask fields. It verifies mask shapes and ownership before
constructing a weighted `DiffusionTokenBatch`. Diffusion evaluation already
returns a globally weighted mean, so the SFT hooks do not divide it by the
evaluation-step count again.

Tunix NNX graph caching remains enabled for normal dense training and disabled
for MoE. It is also disabled for block diffusion when internal neural network
metrics are recorded, because those intermediate graph nodes mutate the cached
graph structure.

## Compatibility

`training_objective=causal_lm` remains the default. Its shifted targets,
MaxText SFT trainer, completion masking, metrics, and dense graph-cache path
are unchanged. Focused tests cover both objective routes and the existing
causal fallback behavior.

## Tests

The current-main replay passed the full CPU gate against the exact upstream
Tunix #1891 integration:

```text
399 passed, 98 skipped, 3 deselected, 219 subtests passed
```

The 98 skips are platform-specific. The three pre-existing Hugging Face
iterator integration tests are deselected locally because they require live
`datasets.load_dataset` access; the changed objective-selection tests in that
module pass.

The gate covers block-diffusion primitives and attention, CFT/SFT corruption,
role masks, prompt-leak rejection, Tunix batch and logits adapters, weighted
metrics, cache routing, configuration, causal regressions, and positive- and
zero-weight gradient accumulation. The Tunix adapter has 100% statement and
branch coverage.

Pyink, repository-equivalent Pylint 3.3.8 (`10.00/10`), Python compilation,
and `git diff --check` also pass. The guarded publisher reruns these
checks on freshly fetched MaxText and Tunix upstream commits before it may
push.

## Checklist

- [x] I performed a self-review of the focused CFT/SFT boundary.
- [x] Existing causal SFT behavior remains the default and has regression coverage.
- [x] Unsupported training modes and unsafe completion layouts fail closed.
- [x] No rollout, distillation, replay, or RL implementation is included.

Design document:
https://docs.google.com/document/d/1N7KcCoAIErB2CV9EJ2G1u_mdN-AQgqwNUYSvM0mtqMI/edit

### CODE REVIEW

File: src/maxtext/integration/tunix/diffusion_sft.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+)
+
+
+def _concrete_numpy(value):
+  """Returns a host view only when eager validation is safe."""
+  if isinstance(value, jax.core.Tracer):
+    return None
+  if isinstance(value, jax.Array) and not value.is_fully_addressable:
+    return None
+  return np.asarray(value)
```
[the remark this answers is not in the document]
Comment (ethannnnnn) [reply src/maxtext/integration/tunix/diffusion_sft.py:46]: Thanks for flagging this. Fixed: `_concrete_numpy` now skips every `jax.Array`, tracer, and `ShapeDtypeStruct`, so device-backed batches no longer enter NumPy validation or trigger per-step device-to-host synchronization. Shape checks still run, while detailed value validation remains enabled for host-resident NumPy batches. Regression tests cover both paths.

File: src/maxtext/trainers/post_train/sft/hooks.py [hunk 2]
```diff
@@ -29,7 +29,15 @@ class SFTTrainingHooks(BaseTrainingHooks):
...
-    return jnp.sum(batch["targets_segmentation"] != 0)
+    loss_weights = batch.get("targets_loss_mask")
+    if loss_weights is None:
+      loss_weights = batch["targets_segmentation"]
+    return jnp.sum(loss_weights != 0)
+
+  @override
+  def eval_loss_is_preaveraged(self) -> bool:
+    """Whether the configured objective supplies a preaveraged eval loss."""
+    return bool(self.config.loss_is_preaveraged or self.config.training_objective == "block_diffusion")
```
Comment (hengtaoguo) [inline src/maxtext/trainers/post_train/sft/hooks.py:40]: I wonder whether it's really necessary to expose `loss_is_preaveraged` to end users (adds up complexity). Can we remove this flag from base/types and instead do this?

```
  @override
  def eval_loss_is_preaveraged(self) -> bool:
    """Whether the configured objective supplies a preaveraged eval loss."""
    return getattr(self.config, "training_objective", "causal_lm") == "block_diffusion"
```

File: src/maxtext/input_pipeline/hf_data_processing.py [hunk 3]
```diff
@@ -48,17 +48,19 @@ def _get_training_objective_transform(
     shift: bool,
     use_dpo: bool,
     use_sft: bool,
+    completion_only: bool,
```
Comment (hengtaoguo) [inline src/maxtext/input_pipeline/hf_data_processing.py:51]: This also changes the calling APIs and may cause incompatibility issues. I wonder if you could directly do `completion_only = bool(use_sft and getattr(config, "sft_train_on_completion_only", False))` inside the `_get_training_objective_transform()` function.

File: src/maxtext/integration/tunix/diffusion_sft.py [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+def create_batch_adapter(config):
+  """Builds a raw MaxText batch adapter for Tunix diffusion SFT."""
+  alignment = config.block_diffusion_logit_alignment
+  block_size = int(config.causal_block_size)
+  completion_only = bool(config.sft_train_on_completion_only)
+
+  def adapt(raw_batch: Mapping[str, Any]) -> diffusion_types.DiffusionTokenBatch:
+    unavailable = sorted(name for name in _REQUIRED_FIELDS if name not in raw_batch or raw_batch[name] is None)
+    if unavailable:
+      raise ValueError("block-diffusion SFT requires non-None batch fields; " f"unavailable {unavailable}")
```
Comment (hengtaoguo) [inline src/maxtext/integration/tunix/diffusion_sft.py:122]: nit

`raise ValueError(f"block-diffusion SFT requires non-None batch fields; unavailable {unavailable}")`

File: src/maxtext/trainers/post_train/sft/hooks.py (hunk 2 above)
Comment (ethannnnnn) [reply src/maxtext/trainers/post_train/sft/hooks.py]: Agreed. I removed the user-facing `loss_is_preaveraged` setting. `SFTTrainingHooks.eval_loss_is_preaveraged()` now derives this directly from `training_objective == "block_diffusion"`, with tests for both causal and diffusion objectives.

File: src/maxtext/input_pipeline/hf_data_processing.py (hunk 3 above)
Comment (ethannnnnn) [reply src/maxtext/input_pipeline/hf_data_processing.py]: Agreed. I removed the `completion_only` argument from `_get_training_objective_transform()` and now derive it inside the helper from `use_sft` and `config.sft_train_on_completion_only`. Tests cover both configured values.

File: src/maxtext/integration/tunix/diffusion_sft.py (hunk 4 above)
Comment (ethannnnnn) [reply src/maxtext/integration/tunix/diffusion_sft.py]: Done. The error message now uses a single f-string as suggested.
